// node-sql-parser/lib/union.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "./create", "./alter", "./select", "./delete", "./update", "./insert", "./command", "./expr", "./limit", "./with", "./util"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("./create"), require("./alter"), require("./select"), require("./delete"), require("./update"), require("./insert"), require("./command"), require("./expr"), require("./limit"), require("./with"), require("./util"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.create, global.alter, global.select, global._delete, global.update, global.insert, global.command, global.expr, global.limit, global._with, global.util);
    global.union = mod.exports;
  }
})(typeof globalThis !== "undefined" ? globalThis : typeof self !== "undefined" ? self : this, function (_exports, _create, _alter, _select, _delete, _update, _insert, _command, _expr, _limit2, _with, _util) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.bigQueryToSQL = bigQueryToSQL;
  _exports.unionToSQL = unionToSQL;
  _exports.multipleToSQL = multipleToSQL;
  const typeToSQLFn = {
    alter: _alter.alterToSQL,
    create: _create.createToSQL,
    select: _select.selectToSQL,
    delete: _delete.deleteToSQL,
    update: _update.updateToSQL,
    insert: _insert.insertToSQL,
    drop: _command.commonCmdToSQL,
    truncate: _command.commonCmdToSQL,
    declare: _command.declareToSQL,
    use: _command.useToSQL,
    rename: _command.renameToSQL,
    call: _command.callToSQL,
    set: _command.setVarToSQL,
    lock: _command.lockUnlockToSQL,
    unlock: _command.lockUnlockToSQL
  };

  function unionToSQL(stmt) {
    const fun = typeToSQLFn[stmt.type];
    const res = [fun(stmt)];
    const {
      _orderby,
      _limit
    } = stmt;

    while (stmt._next) {
      const unionKeyword = (stmt.union || 'union').toUpperCase();
      res.push(unionKeyword, fun(stmt._next));
      stmt = stmt._next;
    }

    res.push((0, _expr.orderOrPartitionByToSQL)(_orderby, 'order by'), (0, _limit2.limitToSQL)(_limit));
    return res.filter(_util.hasVal).join(' ');
  }

  function bigQueryToSQL(stmt) {
    const {
      with: withExpr,
      parentheses,
      select,
      orderby,
      limit
    } = stmt;
    const result = [(0, _with.withToSQL)(withExpr), parentheses && '(', unionToSQL(select), parentheses && ')']; // process with, orderby and limit

    result.push((0, _expr.orderOrPartitionByToSQL)(orderby, 'order by'), (0, _limit2.limitToSQL)(limit));
    return result.filter(val => val).join(' ');
  }

  function multipleToSQL(stmt) {
    const res = [];

    for (let i = 0, len = stmt.length; i < len; ++i) {
      let astInfo = stmt[i] && stmt[i].ast;
      if (!astInfo) astInfo = stmt[i];
      res.push(unionToSQL(astInfo));
    }

    return res.join(' ; ');
  }
});
